File: docs/source/_static/examples/python/scf_solver.py
"""Complete SCF workflow example with settings."""

# --------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE.txt in the project root for license information.
# --------------------------------------------------------------------------------------------

import numpy as np
from qdk_chemistry.algorithms import available, create
from qdk_chemistry.data import Structure

# Create a Structure (coordinates in Bohr/atomic units)
coords = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.4]])
structure = Structure(coords, ["H", "H"])

# start-cell-1
# List available SCF solver implementations
available_solvers = available("scf_solver")
print(f"Available SCF solvers: {available_solvers}")

# Create the default ScfSolver instance
scf_solver = create("scf_solver")
# end-cell-1

# start-cell-2
# Configure the SCF solver using the settings interface
scf_solver.settings().set("basis_set", "sto-3g")
scf_solver.settings().set("method", "hf")
scf_solver.settings().set("max_iterations", 100)
scf_solver.settings().set("tolerance", 1.0e-8)
# end-cell-2

# start-cell-3
# Run the SCF calculation
E_scf, wfn = scf_solver.run(structure, charge=0, spin_multiplicity=1)
scf_orbitals = wfn.get_orbitals()

print(f"SCF Energy: {E_scf:.10f} Hartree")
print(f"Number of molecular orbitals: {scf_orbitals.get_num_molecular_orbitals()}")
print(f"Is restricted: {scf_orbitals.is_restricted()}")
# end-cell-3

File: docs/source/_static/examples/cpp/scf_solver.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE.txt in the project root for
// license information.

// Scf Solver usage examples.
// --------------------------------------------------------------------------------------------
// start-cell-create
#include <iostream>
#include <qdk/chemistry.hpp>
#include <string>
using namespace qdk::chemistry::algorithms;
using namespace qdk::chemistry::data;

// Create default ScfSolver instance
auto scf_solver = ScfSolverFactory::create();
// end-cell-create
// --------------------------------------------------------------------------------------------

int main() {
  // --------------------------------------------------------------------------------------------
  // start-cell-configure
  // Set the method
  // Note the following line is optional, since hf is the default method
  scf_solver->settings().set("method", "hf");

  // end-cell-configure
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-run
  // Load structure from XYZ file
  auto structure = Structure::from_xyz_file("../data/h2.structure.xyz");

  // Run the SCF calculation
  auto [E_scf, wfn] = scf_solver->run(structure, 0, 1, "def2-tzvpp");
  auto scf_orbitals = wfn->get_orbitals();
  std::cout << "SCF Energy: " << E_scf << " Hartree" << std::endl;
  // end-cell-run
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-alternative-run
  // Run scf with an initial guess from previous orbitals
  auto [E_scf2, wfn2] = scf_solver->run(structure, 0, 1, scf_orbitals);

  // Run scf with a custom basis set
  auto basis_set = BasisSet::from_basis_name("def2-tzvpp", structure);
  auto [E_scf3, wfn3] = scf_solver->run(structure, 0, 1, basis_set);
  // end-cell-alternative-run
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-list-implementations
  auto names = ScfSolverFactory::available();
  for (const auto& name : names) {
    std::cout << name << std::endl;
  }
  // end-cell-list-implementations
  // --------------------------------------------------------------------------------------------

  // --------------------------------------------------------------------------------------------
  // start-cell-dfj
  // Run SCF with density-fitted Coulomb integrals (DF-J)
  // Create a basis set with an auxiliary basis for density fitting
  auto dfj_basis = BasisSet::from_basis_name("def2-svp", "def2-universal-jfit",
                                             structure);

  // Configure the solver to use incore ERIs (required for DF-J)
  auto dfj_solver = ScfSolverFactory::create();
  dfj_solver->settings().set("eri_method", "incore");

  // Run - DF-J is automatically enabled when auxiliary basis is detected
  auto [E_dfj, wfn_dfj] = dfj_solver->run(structure, 0, 1, dfj_basis);
  std::cout << "DF-J SCF Energy: " << E_dfj << " Hartree" << std::endl;
  // end-cell-dfj
  // --------------------------------------------------------------------------------------------

  return 0;
}
